src/rgw/rgw_sal_motr.cc

```diff
@@ -2178,65 +2235,84 @@ int MotrObject::get_bucket_dir_ent(const DoutPrefixProvider *dpp, rgw_bucket_dir
                           << rc << dendl;
         return rc;
       }
-
       iter = bl.cbegin();
       ent.decode(iter);
       rc = 0;
       // Put into the cache
       this->store->get_obj_meta_cache()->put(dpp, this->get_name(), bl);
       goto out;
-
     }
     else
-    {  // Version-id instance is empty
-       // Cache miss.
-        keys[0] = this->get_name();
-
-        // Retrieve all 'max' number of pairs.
-        rc = store->next_query_by_name(bucket_index_iname, keys, vals, this->get_name());
-        if (rc < 0) {
-          ldpp_dout(dpp, 0) << __func__ << "ERROR: NEXT query failed. " << rc << dendl;
-          return rc;
-        }
-
-        rc = -ENOENT;
-
-        // Iterating on object keys and return the latest object version
-        for (const auto& bl: vals) {
-          if (bl.length() == 0)
-            break;
+    {
+      // Version-id instance is empty
+      // Cache miss.
+      keys[0] = this->get_name();
+      // Retrieve all 'max' number of pairs.
+      rc = store->next_query_by_name(bucket_index_iname, keys, vals, this->get_name());
+      if (rc < 0) {
+        ldpp_dout(dpp, 0) << __func__ << "ERROR: NEXT query failed. " << rc << dendl;
+        return rc;
+      }
 
-          iter = bl.cbegin();
-          ent.decode(iter);
+      // Iterating on object keys and return the latest object version
+      int i = 1;
+      for (; i < rc; ++i) {
+        if (vals[i].length() == 0) 
+          break;
 
-          if (ent.is_current()) {
-            ldpp_dout(dpp, 20) <<__func__<< ": found current version!" << dendl;
-            rc = 0;
-            // Put into the cache
-            this->store->get_obj_meta_cache()->put(dpp, this->get_name(), bl);
-            break;
-          }
+        iter = vals[i].cbegin();
+        ent.decode(iter);
+        if (ret_code == 0 && ent_to_check.key.instance == null_key.key.instance)
+          return ret_code;
+
+        if (keys[i] == null_version_key)
+          continue;
+
+        if (ent.is_current()) {
+          ldpp_dout(dpp, 20) <<__func__<< ": found current version!" << dendl;
+          rc = 0;
+          // Put into cache
+          this->store->get_obj_meta_cache()->put(dpp, this->get_name(), bl);
+          break;
         }
-     }
-    } else {
-    if (this->store->get_obj_meta_cache()->get(dpp, this->get_name(), bl)) {
+        this->store->get_obj_meta_cache()->put(dpp, this->get_key().to_str(), bl);
+      }
+    }
+  } else {
+    if (this->store->get_obj_meta_cache()->get(dpp, this->get_key().to_str(), bl)) {
       ldpp_dout(dpp, 20) <<__func__<< ": non-versioned bucket!" << dendl;
 
-      // TODO : Handle null version-id scenarios
-
-      rc = this->store->do_idx_op_by_name(bucket_index_iname,
-                                          M0_IC_GET, this->get_key().to_str(), bl);
+      keys[0] = this->get_name();
```

Comment:
lookup using key_name + '['
We first need update put_object code to start using the version ids in key for the kv store. Otherwise there will be re-work here.

Reply:
done